### PR TITLE
LibJS: Super basic passes for "optimisation", plus basic block merging and placement

### PR DESCRIPTION
**Pull request by alimpfard**
TL;DR: less jumping and some scaffolding for future optimisations.
This PR just does a simple CFG simplification, ultimately.
Example result from:
```js
try {
    switch(1) {
        case 2:
            break;
        case 3:
        default:
            console.log("Hurray!");
    }
} catch(e) {
    console.log("Merge me");
}
```
Which merges 2 sets of 2 blocks:
```
CFG Dump for 11 basic blocks:
10 -> 11
5 -> 7
5 -> 6
8 -> 10
1 -> 4
1 -> 2
9 -> 10
7 -> 8
7 -> 9
2 -> 3
6 -> 11
4 -> 5

CFG Dump for 8 basic blocks:
10 -> 11
merge.4.5 -> 7
merge.4.5 -> 6
1 -> merge.4.5
1 -> merge.2.3
7 -> 10
7 -> 8
6 -> 11
8 -> 10

Optimisation passes took 331us
1:
[   0] EnterUnwindContext handler:@merge.2.3 finalizer:<empty> entry:@merge.4.5
merge.4.5:
[   0] LoadImmediate 1
[  18] Store $5
[  20] LoadImmediate 2
[  38] TypedEquals $5
[  40] JumpConditional true:@6 false:@7
7:
[   0] LoadImmediate 3
[  18] TypedEquals $5
[  20] JumpConditional true:@8 false:@10
10:
[   0] LoadImmediate undefined
[  18] LoadImmediate undefined
[  30] Store $7
[  38] GetVariable 1 (console)
[  48] Store $7
[  50] GetById 2 (log)
[  60] Store $6
[  68] NewString 4 ("Hurray!")
[  78] Store $8
[  80] Call callee:$6, this:$7, arguments:[$8]
[  9c] Jump @11
11:
8:
[   0] LoadImmediate undefined
[  18] Jump @10
6:
[   0] LoadImmediate undefined
[  18] Jump @11
merge.2.3:
[   0] LeaveUnwindContext
[   4] SetVariable 0 (e)
[  14] LoadImmediate undefined
[  2c] Store $3
[  34] GetVariable 1 (console)
[  44] Store $3
[  4c] GetById 2 (log)
[  5c] Store $2
[  64] NewString 3 ("Merge me")
[  74] Store $4
[  7c] Call callee:$2, this:$3, arguments:[$4]
```

### CODE REVIEW

**Comment by alimpfard** [comment]
le "one more thing", verified it works fine with generator functions (which are currently the most complex piece of logic we have here)
It also removes duplicates basic blocks (`switch` likes to generate these).
Source:
```js
function* foo(x, y) {
    for (;;) {
        switch(x) {
            case y:
                y = yield x + y;
                if (x > y);
                    for (let a = x; a < y; ++a)
                        yield a;
                break;
            case 69:
                y = yield 1 + x;
                break;
        }
        y = yield 'Ha! good try';
    }
    throw "NO!";
}
```

Before opts     |    After opts
```
Compiled Bytecode::Block for function 'foo':      Compiled Bytecode::Block for function 'foo':
1:                                                1:
[   0] Yield continuation:@2                      [   0] Yield continuation:@2
2:                                                2:
[   0] LoadImmediate undefined                    [   0] LoadImmediate undefined
[  18] Store $2                                   [  18] Store $2
[  20] Jump @4                                    [  20] Jump @merge.4.5
3:                                                merge.4.5:
[   0] Load $2                                    [   0] GetVariable 0 (x)
[   8] NewString 4 ("NO!")                        [  10] Store $3
[  18] Throw                                      [  18] GetVariable 1 (y)
4:                                                [  28] TypedEquals $3
[   0] GetVariable 0 (x)                          [  30] JumpConditional true:@6 false:@7
[  10] Store $3                                   7:
[  18] Jump @5                                    [   0] LoadImmediate 69
5:                                                [  18] TypedEquals $3
[   0] GetVariable 1 (y)                          [  20] JumpConditional true:@8 false:@9
[  10] TypedEquals $3                             9:
[  18] JumpConditional true:@6 false:@7           [   0] LoadImmediate undefined
6:                                                [  18] Jump @10
[   0] LoadImmediate undefined                    10:
[  18] GetVariable 0 (x)                          [   0] NewString 3 ("Ha! good try")
[  28] Store $4                                   [  10] Yield continuation:@20
[  30] GetVariable 1 (y)                          20:
[  40] Add $4                                     [   0] SetVariable 1 (y)
[  48] Yield continuation:@11                     [  10] Jump @merge.4.5
7:                                                8:
[   0] LoadImmediate 69                           [   0] LoadImmediate undefined
[  18] TypedEquals $3                             [  18] LoadImmediate 1
[  20] JumpConditional true:@8 false:@9           [  30] Store $8
8:                                                [  38] GetVariable 0 (x)
[   0] LoadImmediate undefined                    [  48] Add $8
[  18] LoadImmediate 1                            [  50] Yield continuation:@19
[  30] Store $8                                   19:
[  38] GetVariable 0 (x)                          [   0] SetVariable 1 (y)
[  48] Add $8                                     [  10] Jump @10
[  50] Yield continuation:@19                     6:
9:                                                [   0] LoadImmediate undefined
[   0] LoadImmediate undefined                    [  18] GetVariable 0 (x)
[  18] Jump @10                                   [  28] Store $4
10:                                               [  30] GetVariable 1 (y)
[   0] NewString 3 ("Ha! good try")               [  40] Add $4
[  10] Yield continuation:@20                     [  48] Yield continuation:@merge.11.13
11:                                               merge.11.13:
[   0] SetVariable 1 (y)                          [   0] SetVariable 1 (y)
[  10] GetVariable 0 (x)                          [  10] GetVariable 0 (x)
[  20] Store $5                                   [  20] Store $5
[  28] GetVariable 1 (y)                          [  28] GetVariable 1 (y)
[  38] GreaterThan $5                             [  38] GreaterThan $5
[  40] JumpConditional true:@12 false:@13         [  40] GetVariable 0 (x)
12:                                               [  50] SetVariable 2 (a)
[   0] Jump @13                                   [  60] LoadImmediate undefined
13:                                               [  78] Store $6
[   0] GetVariable 0 (x)                          [  80] Jump @16
[  10] SetVariable 2 (a)                          16:
[  20] LoadImmediate undefined                    [   0] GetVariable 2 (a)
[  38] Store $6                                   [  10] Store $7
[  40] Jump @16                                   [  18] GetVariable 1 (y)
14:                                               [  28] LessThan $7
[   0] Load $6                                    [  30] JumpConditional true:@15 false:@14
[   8] Jump @10                                   15:
15:                                               [   0] GetVariable 2 (a)
[   0] GetVariable 2 (a)                          [  10] Yield continuation:@17
[  10] Yield continuation:@18                     17:
16:                                               [   0] GetVariable 2 (a)
[   0] GetVariable 2 (a)                          [  10] Increment
[  10] Store $7                                   [  14] SetVariable 2 (a)
[  18] GetVariable 1 (y)                          [  24] Jump @16
[  28] LessThan $7                                14:
[  30] JumpConditional true:@15 false:@14         [   0] Load $6
17:                                               [   8] Jump @10
[   0] GetVariable 2 (a)                          3:
[  10] Increment                                  [   0] Load $2
[  14] SetVariable 2 (a)                          [   8] NewString 4 ("NO!")
[  24] Jump @16                                   [  18] Throw
18:                                               
[   0] Jump @17                                   
19:                                               
[   0] SetVariable 1 (y)                          
[  10] Jump @10                                   
20:                                               
[   0] SetVariable 1 (y)                          
[  10] Jump @4                                    
                                                  

```

**Comment by alimpfard** [comment]
I declare this good enough :tm: for now.
It'll be a little bit of a maintenance to handle terminating instructions in the CFG generator, but other than that it's mostly self-contained.
Further bikeshedding / comments are of course, welcomed with open virtual arms.

File: Userland/Libraries/LibJS/Bytecode/Interpreter.cpp [hunk 1]
```diff
@@ -165,4 +165,35 @@ void Interpreter::continue_pending_unwind(Label const& resume_label)
         jump(resume_label);
     }
 }
+
+HashMap<UnderlyingType<Interpreter::OptimizationLevel>, NonnullOwnPtr<PassManager>> Interpreter::s_optimization_pipelines {};
```
Comment (awesomekling) [inline Userland/Libraries/LibJS/Bytecode/Interpreter.cpp:169]: This intuitively feels like it should be an array instead of a hash map.

Comment (alimpfard) [reply Userland/Libraries/LibJS/Bytecode/Interpreter.cpp:169]: Then the entries would all be allocated and initialised, or I'd have to switch to storing OwnPtr's 🤔 
I guess the second one works, though I'd have to add a dummy enum member for that.
Sure, why not

File: Userland/Libraries/LibJS/Bytecode/Instruction.h [hunk 2]
```diff
@@ -80,10 +80,12 @@ class Instruction {
 #undef __BYTECODE_OP
     };
 
+    bool is_terminator() const;
     Type type() const { return m_type; }
     size_t length() const;
     String to_string(Bytecode::Executable const&) const;
     void execute(Bytecode::Interpreter&) const;
+    void replace_references_to(BasicBlock const&, BasicBlock const&);
```
Comment (awesomekling) [inline Userland/Libraries/LibJS/Bytecode/Instruction.h:88]: This declaration is highly ambiguous.
The `_to` suffix in the name makes it sound like the first parameter is the "to" target, which it is not.
The parameters are not named and both have the same type so it's entirely unclear which is which.
How about:
```cpp
void replace_block_references(BasicBlock const& from, BasicBlock const& to);
```

Comment (alimpfard) [reply Userland/Libraries/LibJS/Bytecode/Instruction.h:88]: I mean... We'll also need to replace uses of registers, which is why I kept the name generic (so we could later add a Register overload)
Would removing the `_to` resolve this?

Comment (alimpfard) [reply Userland/Libraries/LibJS/Bytecode/Instruction.h:88]: Oh right, now I see why it was named `replace_references_to`.
classes inherit the function named `replace_references` from `Instruction`, so not implementing it wouldn't be a compile error, just a runtime infinite recursion...

Comment (alimpfard) [reply Userland/Libraries/LibJS/Bytecode/Instruction.h:88]: I believe the same can happen for `execute()`.

**Pull request merged**